samza-api/src/main/java/org/apache/samza/operators/windows/Time.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.operators.windows;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Specifies a time interval for applying a window transform over a {@code MessageStream}. Users should
+ * rely on the factory methods to create instances of this class.
+ *
+ * <pre> {@code
+ *  Time threeDays = Time.of(3, TimeUnit.DAYS);
+ *  Time threeMins = Time.minutes(3);
+ *  Time threeSeconds = Time.of(3, TimeUnit.SECONDS);
+ *  MessageStream<WindowOutput<>> windowed = integerStream.window(Windows.tumblingWindow(threeMins));
+ * }
+ * </pre>
+ */
+public final class Time {
+
+  private final TimeUnit timeUnit;
+
+  private final long value;
+
+  private final Time.TimeCharacteristic characteristic = Time.TimeCharacteristic.PROCESSING_TIME;
+
+  /*
+   * Should this {@link Time} be specified in terms of event time or processing time.
+   */
+  enum TimeCharacteristic {
+    EVENT_TIME, PROCESSING_TIME
+  }
+
+  /**
+   * Private constructor to allow instantiation only via factory methods
+   */
+  private Time(long value, TimeUnit timeunit) {
+    this.timeUnit = timeunit;
+    this.value = value;
+  }
+
+  public TimeUnit getTimeUnit() {
+    return timeUnit;
+  }
+
+  public long getValue() {
+    return value;
+  }
+
+  public long toMilliseconds() {
```

Comment:
Slightly prefer the `toMillis` version, since it is explicit that it's converting a duration `to` milli seconds.
Conversion methods in the JDK (that typically perform some local operation on a field) have a `toXX` signature.  eg: `toString`. I typically use `getXX` for getters to fields usually initialized by other public APIs / constructors.